### PR TITLE
#20: Real User Data and Total Outcomes in Trainee Pod Screen

### LINKED ISSUE #20: Calculate Progress Bars for Trainee Pod's Focus Areas (on Pod Screen)
### Ticket
As a youth, I want to see a progress bar and # of outcomes achieved for each of the Trainee Pod's Focus Areas, so that I can visually and mathematically know what my progress for each Focus Area of the Trainee Pod is.

### Acceptance Criteria 
Given I'm on the Home Screen, when I click on the Trainee Pod, then I see a progress bar under each Focus Area name, and a "x out of y Outcomes Achieved" under each progress bar.

### Technical Specifications
- Backend: For each calculation, use a nominator and denominator. Then scale it to Native Base's progress bar calculation. 

### # of Outcomes for Trainee Pod
* Competencies: **3** total outcomes
* Career Exploration: **2** total outcomes
* Life Essentials & Support Network: **2** total outcomes

### PR DESCRIPTION
**Pull request by soph-yang**
- Added to the outcomes endpoint in app.py to calculate the total number of outcomes and then used that info in the frontend of TraineePod.js for the progress bars
- Change to using real user data instead of fake user data by passing in the user parameter in app.py and by using token in TraineePod.js (also changed from axios to fetch).

**Steps to test PR:** 
- Login in on app using Yichen's account to test having real user data: 
Username: my.waymao@gmail.com
password: ZijReFQ@AJYz3jfrA3iz

- Then after clicking on Trainee Pod from Home Screen, should be able to load all the progress bars with the correct number of total outcomes and completed outcomes.

Can do "print(total_and_completed_outcomes)" in app.py to see all the dictionary with total outcomes info

Standard Checklist:
- [ x ] Does the code work? Does it perform its intended function, the logic is correct etc.
- [ x ] Have you merged the most recent version of staging into your branch?
- [ x ] Are all merge conflicts resolved? 
- [ x ] Can any logging or debugging code be removed? (Any debugging statements should be removed)
- [ x ] Is the `diff` the **minimum amount** needed to achieve this ticket's goal?

### CODE REVIEW

File: backend/app.py [hunk 1]
```diff
@@ -62,17 +62,39 @@ def verify():
...
 
-    # salesforce query of each completed outcome # in trainee pod, based on the email and name
-    outcomes_result = sf.query(format_soql("SELECT TR_CareerExpl_Outcomes__c, TR_Competency_Outcomes__c, TR_LifeEssentials_Outcomes__c FROM Trainee_POD_Map__c WHERE (Contact__r.email = {email_value} AND Contact__r.name={full_name})",
-                email_value = email, full_name=name))   
-    return outcomes_result
+    # salesforce query to calculate total outcomes
+    desc = sf.Trainee_POD_Map__c.describe()
+    field_names_and_labels = [(field['name'], field['label']) for field in desc['fields']]
+    field_names = [field['name'] for field in desc['fields']]
+    soql = "SELECT {} FROM Trainee_POD_Map__c".format(','.join(field_names))
```
Comment (jchin01) [inline backend/app.py]: Part 1 /2

File: backend/app.py [hunk 2]
```diff
@@ -62,17 +62,39 @@ def verify():
...
+            compet_count += 1
+        elif "Outcome_LIF" in name_and_label[0]: 
+            life_count += 1
+        elif "Outcome_CAR" in name_and_label[0]:
+            car_count += 1
+    
+    # salesforce query of each completed outcome # in trainee pod, based on the email and name    
+    total_and_completed_outcomes = sf.query(format_soql("SELECT TR_CareerExpl_Outcomes__c, TR_Competency_Outcomes__c, TR_LifeEssentials_Outcomes__c FROM Trainee_POD_Map__c WHERE (Contact__r.email = {email_value} AND Contact__r.name={full_name})",
+                email_value = email, full_name=name))
+
```
Comment (jchin01) [inline backend/app.py]: Part 2/2
So here you are still hard coding for the focus areas. We would want to be able to fetch all focus areas (even if Noah added new focus areas). You should be getting all focus areas from the desc and then SQL selecting for ~all~ focus areas.
Take a look on how Irina did it
` desc = sf.Trainee_POD_Map__c.describe()
     field_names_and_labels = [(field['name'], field['label']) for field in desc['fields']]
     filtered_field_names = [field for field in field_names_and_labels if "Outcomes__c" in field[0]]
     Trainee_field_names = [field[0] for field in filtered_field_names]`

      # salesforce query of each completed outcome # in trainee pod, based on the email and name		     # salesforce query of each completed outcome # in trainee pod, based on the email and name
`     outcomes_result = sf.query(format_soql("SELECT TR_CareerExpl_Outcomes__c, TR_Competency_Outcomes__c, TR_LifeEssentials_Outcomes__c FROM Trainee_POD_Map__c WHERE (Contact__r.email = {email_value} AND Contact__r.name={full_name})",		     soql = "SELECT {} FROM Trainee_POD_Map__c".format(','.join(Trainee_field_names))
                 email_value = email, full_name=name))   		     sf_result = sf.query(format_soql((soql + " WHERE (Contact__r.email = {email_value} AND Contact__r.name={full_name})"), email_value=email, full_name=name))`

File: backend/app.py [hunk 3]
```diff
@@ -62,17 +62,39 @@ def verify():
...
+    soql = "SELECT {} FROM Trainee_POD_Map__c".format(','.join(field_names))
+    
+    compet_count = life_count = car_count = 0
+    for name_and_label in field_names_and_labels:
+        if "Outcome_COM" in name_and_label[0]: 
+            compet_count += 1
+        elif "Outcome_LIF" in name_and_label[0]: 
+            life_count += 1
+        elif "Outcome_CAR" in name_and_label[0]:
+            car_count += 1
```
Comment (jchin01) [inline backend/app.py]: Similar to the comment below, you hard code check for Outcome_COM, Outcome_LIF, etc.. What would happen in the event that Noah added a focus area? Would you be counting the outcomes in the new focus area?

File: frontend/components/pod_components/TraineePod.js [hunk 4]
```diff
@@ -2,52 +2,57 @@ import React from 'react';
...
-                  life_outcomes: data.records[0].TR_LifeEssentials_Outcomes__c,
-              })
-          })
-          .catch(function (error) {
-              console.log(error);
-          });
+        getAccessToken().then(accessToken => 
+            fetch(server_add + '/calculateProgressBar', {
+                headers: {
+                    "Authorization": "Bearer " + accessToken
```
Comment (jchin01) [inline frontend/components/pod_components/TraineePod.js:38]: Good fetch with auth!

File: frontend/components/pod_components/TraineePod.js [hunk 5]
```diff
@@ -2,52 +2,57 @@ import React from 'react';
...
+     *    - career_total_outcomes: total number of career outcomes
+     *    - life_total_outcomes: total number of life outcomes
      */
     state = {
         compet_outcomes: 0, 
         career_outcomes: 0,
         life_outcomes: 0,
+        compet_total_outcomes: 0,
+        career_total_outcomes: 0,
+        life_total_outcomes: 0,
```
Comment (jchin01) [inline frontend/components/pod_components/TraineePod.js:26]: I would try to think of a way to have as many of these as you fetch. For now this works as intended!

**Review by jchin01 — CHANGES_REQUESTED** [review]
Overall, really good on changing the endpoint to require auth and counting the total number of completed tasks instead of just hardcoding a number! Next steps would be to get rid of all hard coded things in the case that Noah adds a new focus area. We are being a little picky, but it is better to add it now than having to go back to an old project and figure out how to do it. You are really close! Just a few more changes ( i.e take a look at Irinias code) and you'll be there! We are trying to get this done by the end of this sprint and you are quite close!

**Comment by soph-yang** [comment]
I got rid of all hardcoded focus areas in the backend when querying the completed num of outcomes and also when calculating the total num of outcomes. I wanted to ask about how to get rid of hardcoding the outcomes fields in the frontend (TraineePod.js), as I don't have that implemented right now

File: backend/app.py [hunk 6]
```diff
@@ -62,17 +62,36 @@ def verify():
...
+    soql = "SELECT {} FROM Trainee_POD_Map__c".format(','.join(Trainee_field_names))
+    sf_result = sf.query(format_soql((soql + " WHERE (Contact__r.email = {email_value} AND Contact__r.name={full_name})"), email_value=email, full_name=name))
+    
+    # count the *total* outcomes for each field 
+    for field in Trainee_field_names:
+        field_type = field[3:6].upper()
+        sf_result[field_type + "_totalcount"] = 0; #create new value in sf_result dict that will store field's total outcomes 
+        for name_and_label in field_names_and_labels:
+            if "Outcome_" + field_type in name_and_label[0]: 
+                sf_result[field_type + "_totalcount"] += 1
```
Comment (jchin01) [inline backend/app.py]: This is a really good change to be able to count for any focus areas that could be potentially added.

File: backend/app.py [hunk 7]
```diff
@@ -62,17 +62,36 @@ def verify():
...
+    desc = sf.Trainee_POD_Map__c.describe()
+    field_names_and_labels = [(field['name'], field['label']) for field in desc['fields']]
+    
+    # filter to get only the trainee outcome field names 
+    filtered_field_names = [field for field in field_names_and_labels if "Outcomes__c" in field[0]]
+    Trainee_field_names = [field[0] for field in filtered_field_names]
+
+    # salesforce query of each *completed* outcome # in trainee pod, based on the email and name
+    soql = "SELECT {} FROM Trainee_POD_Map__c".format(','.join(Trainee_field_names))
+    sf_result = sf.query(format_soql((soql + " WHERE (Contact__r.email = {email_value} AND Contact__r.name={full_name})"), email_value=email, full_name=name))
```
Comment (jchin01) [inline backend/app.py:84]: nicely done

**Review by jchin01 — COMMENTED** [review]
I will look more into the frontend state setting/ potentially a way we could get around it

**Review by jchin01 — APPROVED** [review]

**Pull request merged**